RPIPs/RPIP-onchain-pdao.md

```diff
@@ -0,0 +1,457 @@
+
+1. **Abstain:** The voter's voting power is contributed to quorum but is neither for nor against the proposal.
+2. **For:** The voter votes in favour of the proposal being executed.
+3. **Against:** The voter votes against the proposal being executed.
+4. **Veto:** The voter votes against the proposal as well as indicating they deem the proposal as spam or malicious. If
+   the veto quorum (as defined by the `proposal.veto.quorum parameter`) is met, the proposal is immediately defeated and
+   the proposer loses their bond. This is to dissuade spam, low quality proposals, or proposals that have not gone
+   through off-chain processes first.
+
+If the sum of voting power of votes exceeds the quorum threshold (as defined by the `proposal.quorum` parameter), the
+proposal SHALL be executed and its changes affected.
+
+To vote on a proposal, a node operator MUST call one of the following methods on `RocketProtocolPropose`:
+
+1. `vote(uint256 proposalID, bool support)` where `support` indicates for or against.
+2. `abstain(uint256 proposalID)` to abstain.
+3. `veto(uint256 proposalID)` to vote against and veto.
+
+### Snapshotting
+
+In order to calculate voting power $P_{n}$ at any given block, the protocol SHALL maintain an on-chain array of changes
+to all input variables to the voting power function. The value at a past block can then be looked up via a
+recency-biased binary search.
+
+The following values SHALL be snapshot each time they are changed on chain:
+
+* RPL staked by a node operator ($S_{x}$)
+* ETH deposited by a node operator ($N_{x}$)
+* ETH taken from the deposit pool by a node operator ($U_{x}$)
+* Delegate of a node operator ($d_{x}$)
+* Price of RPL against ETH ($r$)
+* The number of registered nodes ($N$)
+
+### Proposing
+
+Any node with a non-zero voting power MAY raise a proposal at any time provided they have
+waited `proposal.cooldown.period` since their last proposal. Node operators MUST have an effective RPL stake (minus any
+already locked RPL) greater than the proposal bond (as defined by the `proposal.bond` parameter) at the time of
+proposing. This bond amount SHALL be locked during the course of the proposal. Locked RPL SHALL act the same way as
+regular staked RPL for the purposes of rewards, voting and collateral requirements. A node operator MUST NOT be able to
+withdraw RPL if their staked RPL minus locked RPL is lower than the maximum collateral threshold (currently 150%).
+
+As part of a proposal submission, a node operator MUST provide a merkle pollard across a merkle-sum tree of delegated
+voting power at a block that is at most `proposal.max.block.age` blocks old. From this pollard the protocol SHALL
+calculate the total protocol voting power $P$. The root of this pollard alongside the sum SHALL be stored by the
+protocol against the proposal.
+
+If a proposal is not defeated after `proposal.vote.delay.time` has passed, the proposer MAY unlock their RPL bond and
```

Comment:
I think I may be confused about "defeated" here? Is this "the vote fails", "the vote is vetoed" or "the proposal is challenged and the challenge is lost by the proposer"? I now think it's the last, but have 2 questions:

- Why is this before the concept of challenging
- Don't we need the bond for vetos? Or do we essentially have two slightly different impacts on "veto during vote delay" and "veto after vote delay"?

Reply:
Hmm. You're right "defeated proposal" is overloaded here. I was specifically talking about being defeated by a challenge not a vote here. I will try to reword to make it more clear.

* It's mentioned here to keep it within the proposal section as it's related to proposing.
* Very good point regarding the veto, We will need to hold onto the proposal bond until after the proposal voting period too. I will update the spec to account for this.